### ISSUE TITLE
FL job crash before END_RUN orphans the model and wedges the net queue (should catch + set model ERROR)

### ISSUE THREAD

**Issue by atriaybagur** [londonaicentre/FLIP#727]
## Summary

When an FL job's server-side process dies abnormally (OOM, SIGKILL, any hard crash) **before** NVFLARE fires its `END_RUN` event, the model is never moved to a terminal status. It stays at `TRAINING_STARTED` (or `INITIATED`/`PREPARED`) indefinitely, the net's `FLScheduler` row stays `BUSY`, and **every subsequently queued job on that net is starved forever** — silently. The only recovery today is an operator manually calling `POST /fl/stop/{model_id}`.

We should **detect an orphaned/crashed FL run and set the model to `ERROR`**, which is terminal and frees the net — so a crash for _any_ reason degrades to a failed run, never a wedged queue.

## Observed in production (2026-07-07)

- Model `d7d65959-…` (project `f48850b7-…`), a 2-trust Ark+ fine-tune.
- The `fl-server-net-1` task OOM'd during the post-training **cross-site validation** phase: `MemoryUtilized` ramped to **7929 MiB / 8192 MiB (96.8%)** while broadcasting the ~795 MiB global model to both clients, then the job child process died: `ServerEngine - INFO - Job: … child process exit with return code 1` → `JobRunner - INFO - Try to abort job …`.
- Because the crash preceded `END_RUN`, `PersistToS3AndCleanup` never ran (no results in S3) and the model was **stuck at `TRAINING_STARTED` for ~20 h**.
- `GET /fl/status` reported `net_in_use=false`, both clients `no_jobs` — i.e. the FL net was demonstrably idle, yet the hub still treated the model as actively training.
- The next queued model sat at `INITIATED` and was **never dispatched** until the orphan was manually stopped (`POST /fl/stop/{model_id}` → `STOPPED`), which freed the scheduler and let the queue drain.

## Root cause

1. **Terminal status only flips on `END_RUN`.** `ServerEventHandler.handle_event` (`flip-utils/flip/nvflare/components/flip_server_event_handler.py`) sets the terminal status (`RESULTS_UPLOADED` / `ERROR` / `RESULTS_UPLOAD_FAILED` / `STOPPED`) exclusively in the `EventType.END_RUN` branch. A hard process crash (OOM-kill, segfault, SIGKILL) never delivers `END_RUN`, so the model is left non-terminal with no code path to correct it.

2. **The existing stale-scheduler recovery doesn't cover this case.** `_recover_stale_busy_schedulers` (`flip-api/src/flip_api/fl_services/run_jobs.py:27`) frees a `BUSY` scheduler only when its job is `COMPLETED`/`DELETED` (or `job_id IS NULL`):

   ```sql
   UPDATE fl_scheduler SET status = :available, job_id = NULL
   WHERE status = :busy
     AND (job_id IS NULL
          OR job_id NOT IN (SELECT id FROM fl_job WHERE status NOT IN (:completed, :deleted)))
   ```

   A crashed-but-non-terminal run has the model at `TRAINING_STARTED` and its `fl_job` still `IN_PROGRESS`, so it is treated as legitimately busy and never recovered.

Net effect: a single crash permanently starves the net, with **no error surfaced to the user** — the model UI shows "training" forever.

## Proposed fix

Primary ask — **catch the orphaned run and set the model to `ERROR`** (terminal → frees the scheduler via the existing recovery + surfaces the failure to the user). Suggested layers, in order of value:

1. **Hub-side FL-run watchdog (primary).** In the periodic scheduler (alongside `run_jobs_scheduled_task` / the stale-task recovery in `flip-api/src/flip_api/scheduler/apscheduler_runner.py`), reconcile models in a running state (`INITIATED`/`PREPARED`/`TRAINING_STARTED`) against live FL state from `get_status` (`flip-api/src/flip_api/fl_services/get_status.py`). If the net reports the job gone — fl-server engine not `STARTING`/`STARTED`, or clients `no_jobs`, or `net_in_use=false` — for longer than a grace period (and/or no status/metric heartbeat for _N_ minutes), flip the model to `ERROR`. A per-model "last progress" timestamp/heartbeat would make this robust against slow-but-alive runs (note: a legitimately-slow round-0 broadcast over a slow link can look idle for minutes — the grace period must exceed that).

2. **Server-side abnormal-end report (belt-and-braces).** The fl-server **parent** engine survives the job child crash (the logs show `ServerEngine: child process exit with return code 1` and `JobRunner: Try to abort job`). Have it report `ERROR` to the hub when a job ends without having emitted a terminal FLIP status. This catches the failure faster and closer to the source than the hub watchdog.

3. **Broaden `_recover_stale_busy_schedulers`.** Once the model is terminalized to `ERROR`, the existing recovery frees the net naturally. Optionally also treat a scheduler whose model is non-terminal-but-crashed as recoverable, as a second safety net.

## Impact / severity

- **Availability:** one crash (OOM or otherwise) wedges an entire FL net until manual operator intervention — no self-healing.
- **Silent:** the user sees a perpetual "training" state; no `ERROR`, no notification.
- **Data:** the trained model is also lost (crash precedes persist) — tracked separately.

## Related

- The specific OOM **trigger** here (post-training cross-site-validation broadcasting the full model) is being addressed separately by switching finetune post-training eval to `GlobalModelEval` (+ head-only eval). **This issue is deliberately scoped to fault-tolerance:** a crash for _any_ reason must degrade to `ERROR`, not a silently-wedged queue.

**Comment by garciadias** [comment]
## Independent confirmation (2026-07-07, prod)

Investigated this exact model/incident before spotting this issue was already filed. Confirms the write-up above end-to-end from CloudWatch, and adds two details:

**Memory timeline (ECS/ContainerInsights, `fl-server-net-1`, 2 vCPU / 8192 MiB task):**

| Time (UTC) | MemoryUtilized |
|---|---|
| 16:19 | 7157 MiB |
| 16:20 | **7929 MiB (96.8%)** |
| 16:21 | 7929 MiB (96.8%) |
| 16:22 | 5419 MiB (crash) |
| 16:23+ | 434 MiB (idle baseline) |

Matches the issue's numbers exactly.

**The NaN metric errors are a symptom, not a second cause.** Right at 16:22:21 UTC — inside the same second the crash sequence starts — the server logged 3x:
```
FLIP - ERROR - Something went wrong when sending metrics, see exception below
FLIP - ERROR - Out of range float values are not JSON compliant: nan
```
followed immediately by repeated `CoreCell - ERROR` stream failures talking to `Trust_2` (`download_service` stream), then `ServerEngine - INFO - Job: eeb95e41-... child process exit with return code 1` at 16:22:27.927, then `JobRunner: Try to abort job ... on clients` at 16:22:28.012. Worth flagging so nobody chases the NaN as a separate root cause — it's almost certainly the aggregated cross-site-eval metric coming back `nan` because the process was already OOM-strained, not an independent trigger. `flip.send_metrics` already has a broad `try/except` (`flip-utils/flip/core/standard.py:339-357`) so it can't be what tore down the job on its own — the crash is the child-process OOM kill, exactly as diagnosed above.

**Confirms the "manual stop masks the failure" gap in production, live:** model `d7d65959-3e40-4744-b853-0ece208df840` is currently sitting at `ModelStatus.STOPPED` (not `ERROR`) — flip-api logged `No running FL job for model d7d65959-...` at the moment of the manual stop (2026-07-07 11:27:55 UTC), then `stop_training` unconditionally set `STOPPED` per `flip-api/src/flip_api/fl_services/stop_training.py:59-61`, exactly the gap this issue describes. `fl-server-net-1`/`fl-api-net-1` were also force-redeployed ~10:34–11:38 UTC same day as part of the manual recovery, on top of the `POST /fl/stop` call.

No code changes proposed here beyond what's already scoped above — this is corroboration + the "don't chase the NaN" note for whoever picks this up.
